homework02/allan/string.cc

```diff
@@ -0,0 +1,103 @@
+#include "string.h"
+
+#include <cstring>
+#include <iostream>
+String::String(const char *ch)
+  : String(ch, strlen(ch) + 1)
+{
+}
+
+String::String()
+   :String("", 1)
+{
+}
+
+String::String(const String& other)
+  : String(other.c_str(), other.capacity())
+{
+}
+
+
+String::String(const char *ch, int capacity)
+  : mCapacity(capacity),
+    mData(std::make_unique<char []>(mCapacity))
+{
+  strcpy(mData.get(), ch);
+}
+
+String::~String()
+{
+}
+
+String& String::operator=(String other)
+{
+  std::swap(mCapacity, other.mCapacity);
+        std::swap(mData, other.mData);
+}
+
+bool String::empty() const
+{
+        return size() == 0;
+}
+
+const char* String::c_str() const
+{
+        return mData.get();
+}
+
+String::operator std::string() const
+{
+        return std::string(c_str());
+}
+
+char& String::operator[](int pos)
+{
+  return mData[pos];
+}
+
+const char& String::operator[](int pos) const
+{
+  return mData[pos];
+}
+
+int String::size() const
+{
+        return strlen(mData.get());
```

Comment:
Why not - again, its a choice, although I must admit that most string implementations follow a constant complexity when it comes to `size()`, as it is called very often, as  `empty()` most of the time is implemented as `size() == 0`.